ouroboros-consensus/src/Ouroboros/Storage/ChainDB/ImmDB.hs

```diff
@@ -0,0 +1,357 @@
+    case epochOrSlot of
+      Left epochNo -> fmap snd <$> ImmDB.getEBB imm epochNo
+      Right slotNo -> ImmDB.getBinaryBlob imm slotNo
+
+{-------------------------------------------------------------------------------
+  Streaming
+-------------------------------------------------------------------------------}
+
+-- | Stream blocks after the given point
+--
+-- See also 'streamBlobsAfter'.
+streamBlocksAfter :: forall m blk. (MonadCatch m, HasHeader blk, Typeable blk)
+                  => ImmDB m blk
+                  -> Point blk -- ^ Exclusive lower bound
+                  -> m (Iterator (HeaderHash blk) m blk)
+streamBlocksAfter db low = wrap <$> streamBlobsAfter db low
+  where
+    wrap :: Iterator (HeaderHash blk) m Strict.ByteString
+         -> Iterator (HeaderHash blk) m blk
+    wrap itr = Iterator {
+          iteratorNext  = parseIteratorResult db =<< iteratorNext itr
+        , iteratorClose = iteratorClose itr
+        , iteratorID    = ImmDB.DerivedIteratorID $ iteratorID itr
+        }
+
+
+-- | Stream blobs after the given point
+--
+-- Our 'Point' based API and the hash based API of the 'ImmutableDB' have a
+-- slight impedance mismatch: the bounds offered by the 'ImmutableDB' are
+-- inclusive, /if specified/. We can't offer that here, since it does not make
+-- sense to stream /from/ (inclusive) the genesis block. Therefore we provide
+-- an /exclusive/ lower bound as a 'Point': if the 'Point' refers to the genesis
+-- block we don't give the 'ImmutableDB' a lower bound at all; if it doesn't,
+-- we pass the hash as the lower bound to the 'ImmutableDB' and then step the
+-- iterator one block to skip that first block.
+--
+-- If we do skip a block, we return it separately. We don't check the hash,
+-- since we don't deserialize the block. We do this in 'streamBlocksAfter'.
+streamBlobsAfter :: forall m blk. (MonadCatch m, HasHeader blk, Typeable blk)
+                 => ImmDB m blk
+                 -> Point blk -- ^ Exclusive lower bound
+                 -> m (Iterator (HeaderHash blk) m Strict.ByteString)
+streamBlobsAfter db low = withDB db $ \imm -> do
+    itr   <- ImmDB.streamBinaryBlobs imm low' Nothing
+    skipAndCheck itr
+    return itr
+  where
+    low' :: Maybe (SlotNo, HeaderHash blk)
+    low' = case pointHash low of
+             GenesisHash -> Nothing
+             BlockHash h -> Just (pointSlot low, h)
+
+    -- Skip the first block (if any) to provide an /exclusive/ lower bound
+    --
+    -- Take advantage of the opportunity to also verify the hash (which the
+    -- immutable DB can't, since it cannot compute hashes)
+    skipAndCheck :: Iterator (HeaderHash blk) m Strict.ByteString -> m ()
+    skipAndCheck itr =
+        case low' of
+          Nothing           -> return ()
+          Just (slot, hash) -> do
+            skipped <- parseIteratorResult db =<< iteratorNext itr
+            case skipped of
+              IteratorResult slot' blk ->
```

Comment:
Intention here: do some sanity checking. Check the hash, check that the iterator is not already empty, that sort of thing.

Reply:
Can use `ChainDbFailure` (may need to add another constructor).